### DISCUSSION TITLE
In node express with socket.io, why does my Content-Type header not include a charset value when loading the socket.io.js file?

### DISCUSSION THREAD

**Discussion by owntheweb** [socketio/socket.io#4589] [Q&A] [answered]
(Also posted [here](https://stackoverflow.com/questions/75072246/in-node-express-with-socket-io-why-does-my-content-type-header-not-include-a-ch) for internet points)

Hello socket.io gurus,

I have a node express server setup with socket.io. I had a question come up about why charset wasn't being set to UTF-8 in my response header when loading socket.io.js from my client-side html page. If I were to load it [in the socket.io demo page](https://socket.io/demos/chat/), I see the expected charset:

```
content-type: application/javascript; charset=utf-8
```

In my app however, I see:

```
content-type: application/javascript;
```

(note: I need to do a hard refresh to see these headers show up in dev tools otherwise I just get a 304 no change)

Everything else in the app returns the charset header of UTF-8, so this leaves me scratching my head. Here's the minimal version of the app that can reproduce the issue:

help.ts
```
import * as http from 'http'
import express, { Application, Request, Response } from 'express'
import cors from 'cors'
import { Server } from 'socket.io'

let server

const startExpress = (): Promise<http.Server> => {
  return new Promise((resolve, reject) => {
    const app = express()
    app.use(cors())
    const port = Number(process.env.PORT || 3000)

    const server = app
      .listen(port, () => {
        setRoutes(app)
        resolve(server)
      })
      .on('error', (error: Error) => {
        reject(error)
      })
  })
}

const setRoutes = (app: Application) => {
  app.get('/hi', (request: Request, response: Response) => {
    response.sendFile(__dirname + '/hi.html')
  })
}

const setupSocketIoServer = (expressServer: http.Server): void => {
  server = new Server(expressServer, { cors: { origin: '*' } })
  // add events and such, not needed for this demo
}

const go = async () => {
  // start app
  const expressServer = await startExpress()

  // start socket.io
  setupSocketIoServer(expressServer)
}

go()
```

hi.html
```
<!DOCTYPE html>
<html>
  <head>
    <title>Hello</title>
    <script src="/socket.io/socket.io.js"></script>
  </head>
  <body>
    <h1>Hello</h1>
  </body>
</html>
```

Getting it started:
```
npx ts-node help.ts
```

Since this is a module-provided file, I cannot seem to override it with something like:

```
app.use((req, res, next) => {
  if (/.*\.js/.test(req.path)) {
    res.charset = 'utf-8'
  }
  next()
})
```

```
"express": "^4.18.2",
"socket.io": "^4.5.4"
```

I'm stumped. Any idea of what I might be missing?

**Comment by darrachequesne** [comment, marked as answer]
Hi! The charset isn't currently included when serving the client bundle: https://github.com/socketio/socket.io/blob/54d5ee05a684371191e207b8089f09fc24eb5107/lib/index.ts#L567-L570

I guess it's automatically in the chat demo (coming from https://cdnjs.cloudflare.com/ajax/libs/socket.io/2.3.0/socket.io.js).

**Reply by owntheweb** [reply]
Thank you!

**Reply by darrachequesne** [reply]
@ChrisAtBombBomb this will be fixed in next release: https://github.com/socketio/socket.io/commit/93d446a5453894a1e90499856ca731ac397f1275

**Reply by qiulang** [reply]
@darrachequesne  [RFC 9239 Updates to ECMAScript Media Types](https://www.rfc-editor.org/rfc/rfc9239.html#name-common-javascript-media-typ), which was published in May 2022, the Encoding section says

> Module goal sources MUST be encoded as UTF-8; all other encodings will fail. Source goal sources SHOULD be encoded as UTF-8

So does it really necessary to set UTF-8 in source code ?

**Reply by darrachequesne** [reply]
@qiulang I guess it doesn't hurt to set an explicit charset.

@ChrisAtBombBomb https://github.com/socketio/socket.io/commit/93d446a5453894a1e90499856ca731ac397f1275 was included in version [4.6.0](https://github.com/socketio/socket.io/releases/4.6.0) :rocket:
